src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/HashProviderDispenser.Windows.cs

```diff
@@ -24,12 +29,86 @@ public static HashProvider CreateMacProvider(string hashAlgorithmId, ReadOnlySpa
 
         public static class OneShotHashProvider
         {
+            private static volatile bool s_UseCompatOneShot;
+
             public static unsafe int HashData(string hashAlgorithmId, ReadOnlySpan<byte> source, Span<byte> destination)
             {
-                using (HashProviderCng hashProvider = new HashProviderCng(hashAlgorithmId, null))
+                // Shared handle, no using or dispose.
+                SafeBCryptAlgorithmHandle cachedAlgorithmHandle = BCryptAlgorithmCache.GetCachedBCryptAlgorithmHandle(
+                    hashAlgorithmId,
+                    BCryptOpenAlgorithmProviderFlags.None);
+
+                int hashSize;
+
+                NTSTATUS ntStatus = Interop.BCrypt.BCryptGetProperty(
+                    cachedAlgorithmHandle,
+                    Interop.BCrypt.BCryptPropertyStrings.BCRYPT_HASH_LENGTH,
+                    &hashSize,
+                    sizeof(int),
+                    out _,
+                    0);
+
+                if (ntStatus != NTSTATUS.STATUS_SUCCESS)
+                    throw Interop.BCrypt.CreateCryptographicException(ntStatus);
+
+                if (destination.Length < hashSize)
+                    throw new CryptographicException();
+
+                if (!s_UseCompatOneShot)
                 {
-                    hashProvider.AppendHashData(source);
-                    return hashProvider.FinalizeHashAndReset(destination);
+                    try
+                    {
+                        fixed (byte* pSource = source)
+                        fixed (byte* pDestination = destination)
+                        {
+                            ntStatus = Interop.BCrypt.BCryptHash(cachedAlgorithmHandle, null, 0, pSource, source.Length, pDestination, hashSize);
+
+                            if (ntStatus != NTSTATUS.STATUS_SUCCESS)
+                            {
+                                throw Interop.BCrypt.CreateCryptographicException(ntStatus);
+                            }
+                        }
+                    }
```

Comment:
I'd probably put a return after the if-not-success-throw block so that I could avoid having two calls to HashUpdateAndFinish; but it's small enough and close enough together that they probably won't accidentally diverge.